src/main/java/org/zalando/nakadi/validation/schema/SchemaDiff.java

```diff
@@ -0,0 +1,331 @@
+package org.zalando.nakadi.validation.schema;
+
+import org.everit.json.schema.ArraySchema;
+import org.everit.json.schema.CombinedSchema;
+import org.everit.json.schema.EnumSchema;
+import org.everit.json.schema.NumberSchema;
+import org.everit.json.schema.ObjectSchema;
+import org.everit.json.schema.ReferenceSchema;
+import org.everit.json.schema.Schema;
+import org.everit.json.schema.StringSchema;
+import org.zalando.nakadi.domain.SchemaChange;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.Stack;
+import java.util.stream.Collectors;
+
+import static org.everit.json.schema.CombinedSchema.ALL_CRITERION;
+import static org.everit.json.schema.CombinedSchema.ANY_CRITERION;
+import static org.zalando.nakadi.domain.SchemaChange.Type.ADDITIONAL_ITEMS_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.ADDITIONAL_PROPERTIES_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.ATTRIBUTE_VALUE_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.COMPOSITION_METHOD_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.DEPENDENCY_ARRAY_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.DEPENDENCY_SCHEMA_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.DEPENDENCY_SCHEMA_REMOVED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.DESCRIPTION_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.ENUM_ARRAY_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.ID_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.NUMBER_OF_ITEMS_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.PROPERTIES_ADDED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.PROPERTY_REMOVED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.REQUIRED_ARRAY_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.SCHEMA_REMOVED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.SUB_SCHEMA_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.TITLE_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.TYPE_CHANGED;
+
+public class SchemaDiff {
+    public List<SchemaChange> collectChanges(final Schema original, final Schema update) {
+        final List<SchemaChange> changes = new ArrayList<>();
+        final Stack<String> jsonPath = new Stack<>();
+
+        recursiveCheck(original, update, jsonPath, changes);
+
+        return changes;
+    }
+
+    private void recursiveCheck(
+            final Schema original,
+            final Schema update,
+            final Stack<String> jsonPath,
+            final List<SchemaChange> changes) {
+
+        if (original == null && update == null) {
+            return;
+        }
+
+        if (original == null || update == null) {
+            addChange(SCHEMA_REMOVED, jsonPath, changes);
```

Comment:
Actually, I could not think of a scenario when the first `original == null` would happen. Changed here 1d74868